Multi-shell multi-tissue constrained spherical deconvolution

karanphil opened the pull request:
This pull request focuses on multi-shell multi-tissue (msmt) CSD. In a nutshell, before my changes, 

- I wasn't able to produce good white matter fodf's results from msmt-csd using dipy.reconst.mcsd tools. 

- The previous results were having lots of problems distinguishing grey matter properly. 

- There was no function for estimating tissue responses (diffusivities and S0) in msmt and the function generating fiber response functions was incorrect.

 In order to make it all work, here are the changes I made. 

1) First, I took a small detour into the single-shell single-tissue (ssst) code to reshape a little bit the functions used to estimate response functions. 

- Previously, there were auto_response and response_from_mask in dipy.reconst.csdeconv. 

- Now, there are auto_response_ssst, response_from_mask_ssst and mask_for_response_ssst. The latter is a new function creating a mask of voxels that will be used to estimate the response function. This is simply the part of auto_response that  was not the same as response_from_mask. 

- So now auto_response_ssst only calls mask_for_response_ssst, followed by response_from_mask_ssst, which was only renamed.

2) Secondly, I added similar tools to dipy.reconst.mcsd, as auto_response_msmt, response_from_mask_msmt and mask_for_response_msmt. The structure is the same, except it calculates responses for each tissue and each b-value. The use of mask_for_response_msmt followed by response_from_mask_msmt allows to apply tissue segmentation masks between the processes to greatly improve the final masks used to estimate the responses. 

- In order to add these functions, I also had to add new small functions in dipy.core.gradients. 

- The already existing function unique_bvals was renamed unique_bvals_magnitude, as it is using magnitude to find unique b-values. 

- I added a unique_bvals_tolerance function based on a tolerance gap instead. 

- I also added a function called get_bval_indices that allows you to find these indices at which a given b-value was in the complete b-values list. 

- Note that in order to cope well with the ROIs in mask_for_response_msmt, I wrote two functions in dipy.reconst.utils, _roi_in_volume and _mask_in_roi.

3) Thirdly, the multi_shell_fiber_response function has been moved from dipy.sims.voxel to dipy.reconst.mcsd. 

- Then, I found that there was a fundamental error in it and fixed it. Without going into too much detail, it had to do with taking S0 into account and using a different response function for each b-value (I can explain more if needed).

4) Fourthly and finally, I did some modifications in the classes of dipy.reconst.mcsd, namely MultiShellResponse (simply saving S0 as well), MultiShellDeconvModel (calling the right fiber response function and working on the predict function) and MSDeconvFit (adding a way to get the compartments fodf’s).

There are adapted tests for each new or modified function and I explained how to use the new stuff in doc.examples.reconst_mcsd.py. 

Here is a real data example of the white matter fodf’s, before, after and using Mrtrix. We can see that before this PR, there was no grey matter (in green) in the volume fractions map. Now, with this PR, the grey matter is well defined, as well as white matter fodf’s.

![image](https://user-images.githubusercontent.com/47002808/86606927-a005c400-bf76-11ea-9e43-d2bb473522cc.png)

pep8speaks commented:
Hello @karanphil, Thank you for submitting the Pull Request !







* In the file [`dipy/reconst/csdeconv.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/csdeconv.py):

> [Line 28:1](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/csdeconv.py#L28): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 0
> [Line 33:1](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/csdeconv.py#L33): [E304](https://duckduckgo.com/?q=pep8%20E304) blank lines found after function decorator

* In the file [`dipy/reconst/mcsd.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/mcsd.py):

> [Line 463:81](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/mcsd.py#L463): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)



* In the file [`dipy/reconst/tests/test_csdeconv.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_csdeconv.py):

> [Line 61:56](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_csdeconv.py#L61): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 62:56](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_csdeconv.py#L62): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 63:56](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_csdeconv.py#L63): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 64:56](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_csdeconv.py#L64): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 150:43](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_csdeconv.py#L150): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 151:43](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_csdeconv.py#L151): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 152:43](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_csdeconv.py#L152): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

* In the file [`dipy/reconst/tests/test_mcsd.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py):

> [Line 110:81](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L110): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 80 characters)
> [Line 111:81](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L111): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 80 characters)
> [Line 182:5](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L182): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 194:28](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L194): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 197:37](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L197): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 198:37](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L198): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 199:37](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L199): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 199:70](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L199): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 200:37](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L200): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 215:37](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L215): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 216:37](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L216): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 217:37](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L217): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 217:70](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L217): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 218:37](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L218): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 229:41](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L229): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 230:41](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L230): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 231:41](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L231): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 259:49](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L259): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 264:81](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L264): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)
> [Line 287:29](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L287): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 288:29](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L288): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 289:29](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L289): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 298:41](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L298): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 299:41](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L299): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 300:41](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L300): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 300:74](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L300): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 301:41](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/dipy/reconst/tests/test_mcsd.py#L301): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent











* In the file [`doc/examples/contextual_enhancement.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/contextual_enhancement.py):

> [Line 120:1](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/contextual_enhancement.py#L120): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file

* In the file [`doc/examples/fiber_to_bundle_coherence.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/fiber_to_bundle_coherence.py):

> [Line 111:1](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/fiber_to_bundle_coherence.py#L111): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file



* In the file [`doc/examples/reconst_csd.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/reconst_csd.py):

> [Line 62:1](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/reconst_csd.py#L62): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file

* In the file [`doc/examples/reconst_csd_parallel.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/reconst_csd_parallel.py):

> [Line 34:1](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/reconst_csd_parallel.py#L34): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file

* In the file [`doc/examples/reconst_mcsd.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/reconst_mcsd.py):

> [Line 197:74](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/reconst_mcsd.py#L197): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 274:81](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/reconst_mcsd.py#L274): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)

* In the file [`doc/examples/sfm_reconst.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/sfm_reconst.py):

> [Line 83:1](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/sfm_reconst.py#L83): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 91:60](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/sfm_reconst.py#L91): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace





* In the file [`doc/examples/tracking_introduction_eudx.py`](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/tracking_introduction_eudx.py):

> [Line 65:1](https://github.com/dipy/dipy/blob/9921e9b612d200bbaf217f4b399c135c4c96e0d3/doc/examples/tracking_introduction_eudx.py#L65): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file









Do see the [DIPY coding Style guideline](https://github.com/dipy/dipy/blob/master/doc/devel/coding_style_guideline.rst)

karanphil commented:
The workflows/tests/test_reconst_mapmri.py test seems to fail. I don't think it is my fault, but ultimately it could come from a small modification to auto_response_ssst (auto_response). On my computer, it is working well with my changes. Is this normal @skoudro?

skoudoro commented:
> Is this normal @skoudro?

Yes, it is normal, it is fixed in #2188 


Sorry for getting back late to you and thank you for this awesome PR! The results look good. However, We will need time to review it because there is a lot of changes. 

@ShreyasFadnavis, Can you start to have a first look?

frheault commented on dipy/reconst/csdeconv.py:
```diff
@@ -13,15 +14,60 @@
+    Notes
+    -----
+    In CSD there is an important pre-processing step: the estimation of the
+    fiber response function. In order to do this we look for voxels with very
```
In order to do this**,**
(Or simply: _To do this,_)

frheault commented on dipy/reconst/tests/test_mcsd.py:
```diff
@@ -58,42 +91,235 @@ def test_mcsd_model_delta():
 
 
 @pytest.mark.skipif(not mcsd.have_cvxpy, reason="Requires CVXPY")
```
There is a PR #2188 related to mapmri and CVXPY, which is one of the workflow test that are failing, maybe it is related.
(Just so you know, maybe it is not your fault)

frheault commented on dipy/reconst/tests/test_mcsd.py:
```diff
@@ -58,42 +91,235 @@ def test_mcsd_model_delta():
+                                          csf_response)
+    model_1 = MultiShellDeconvModel(gtab, response, sh_order=sh_order)
+    responses = np.array([wm_response, gm_response, csf_response])
+    print(responses.shape)
```
I guess printing is not necessary

frheault commented on doc/examples/reconst_mcsd.py:
```diff
@@ -62,6 +67,8 @@
 bvals, bvecs = read_bvals_bvecs(fbval, fbvec)
 gtab = gradient_table(bvals, bvecs)
 
+data *= 1
```
Why ?

frheault commented on doc/examples/reconst_mcsd.py:
```diff
@@ -162,68 +170,130 @@
+"""
+
+response = np.array([response_wm, response_gm, response_csf])
+rapid_model = MultiShellDeconvModel(gtab, response, sh_order=8)
```
rapid_model ? Weird variable name

frheault commented on doc/examples/reconst_mcsd.py:
```diff
@@ -233,16 +303,42 @@
 mcsd_odf = mcsd_fit.odf(sphere)
-fodf_spheres = actor.odf_slicer(mcsd_odf, sphere=sphere, scale=0.01,
+
+print("ODF")
```
Just in case, are these 3 prints still useful?

frheault commented:
@skoudoro I see that you put this PR in 1.3, but I think it is important to mention that improvements made to the code are not an enhancement or a feature, they are more like a 'bug fix'. The earlier version of the code was not giving correct results for the GM fODF and the structure of the code was not designed to make it (correctly) usable by users.

Considering the test coverage is pretty good and that examples/docs are already fixed, I think it would be nice to include it in 1.2 (unless 1.2 is coming very soon). I don't know when 1.3 is planned for, since 1.2 got delayed does 1.3 get pushed by as much (I know you were trying to have smaller release)

Garyfallidis commented:
@karanphil apologies for any delays. This indeed belongs to the bug fix category as @frheault suggests. @ShreyasFadnavis can you have a look at this PR?

ShreyasFadnavis commented:
Hi @karanphil! Thank you for this much-needed PR 👍 I tested it and looks good. Since we are at it, can we also try to have a non-negativity constraint on the ODFs? This is something in my `TODOs` when the previous PR was merged. Not sure how important that feature is, but could be easy to add in this PR (or the next). Also, Travis seems unhappy due to some unrelated failure which @skoudoro seems to have fixed in #2188. This PR may therefore need rebasing! Overall LGTM and excellent work!

karanphil commented:
Thanks @ShreyasFadnavis for the review! I think adding a non-negativy constraint is a good idea. However, I would put it on a TODO list for a futur PR. I must say that this part of the code is not of my expertise, so maybe we could have a meeting with @mdesco and others about it. What do you think?

karanphil commented:
@ShreyasFadnavis, about the rebasing, I tried rebasing to upstream/master but I have some very old commits that conflict with the master, commits that aren't even up to date with my actual code. Can I juste pull upstream/master into my branch and resolve the conflicts? Or squash my commits and then rebase? Thanks in advance!

skoudoro commented on dipy/core/gradients.py:
```diff
@@ -674,7 +674,87 @@ def round_bvals(bvals, bmag=None):
+    """
+    b = np.unique(bvals)
+    ubvals = []
+    i = 0
+    lower_part = np.where(b <= b[i] + tol)[0]
```
For clarity, Why you do not put directly `b[0]`? What is the goal to initialize `i` here?

ShreyasFadnavis commented:
> Thanks @ShreyasFadnavis for the review! I think adding a non-negativy constraint is a good idea. However, I would put it on a TODO list for a futur PR. I must say that this part of the code is not of my expertise, so maybe we could have a meeting with @mdesco and others about it. What do you think?

Yes, that can be done on a separate PR for sure!

ShreyasFadnavis commented:
> @ShreyasFadnavis, about the rebasing, I tried rebasing to upstream/master but I have some very old commits that conflict with the master, commits that aren't even up to date with my actual code. Can I just pull upstream/master into my branch and resolve the conflicts? Or squash my commits and then rebase? Thanks in advance!

I will defer to @skoudoro on this one! @skoudoro what do you think? Maybe we don't even need to rebase! Should restarting the CI be sufficient?

skoudoro commented on dipy/core/gradients.py:
```diff
@@ -674,7 +674,87 @@ def round_bvals(bvals, bmag=None):
+                                         b > b[lower_part[-1]]))[0]
+    ubvals.append(b[lower_part[-1]])
+    if len(upper_part) != 0:
+        i = upper_part[-1] + 1
+    else:
+        i = lower_part[-1] + 1
+    while i != len(b):
```
What is `i`? Can you rename this variable. It is confusing, specially with the for loop below and the `i=0` initialization above

karanphil replied:
Is `b_index` a better name?

skoudoro replied:
yes

skoudoro commented on dipy/core/gradients.py:
```diff
@@ -674,7 +674,87 @@ def round_bvals(bvals, bmag=None):
+            i = lower_part[-1] + 1
+
+    # Checking for overlap with get_bval_indices
+    for i in range(len(ubvals)-1):
```
We recommend a more pythonic way to do this loop. (`for i, val in enumerate(ubvals[:-1])` for example)

karanphil replied:
I didn't use enumerate because I need ubvals[i] and ubvals[i+1]. Is it ok?

skoudoro replied:
with `enumerate(ubvals[1:])`:
- `val` will be `ubvals[i+1]`
- `ubvals[i]` stay the same
if I am not wrong. 

It is not pythonic to use `range(len())` so we try to avoid this practice.

skoudoro commented on dipy/core/gradients.py:
```diff
@@ -674,7 +674,87 @@ def round_bvals(bvals, bmag=None):
+        if len(np.intersect1d(indices_1, indices_2)) != 0:
+            msg = '''There is overlap in clustering of b-values.
+            The tolerance factor might be too high.'''
+            warn(msg, UserWarning)
```
is it ok to have an overlap of clustering? Why don't you raise an error?

karanphil replied:
It could result in problems for the user, but I didn't know if it could still be used in some cases. I can raise an error.

ShreyasFadnavis replied:
IMO: I think a warning is okay here!

skoudoro replied:
ok, thanks for the explanation @karanphil

skoudoro replied:
you can keep it

skoudoro commented on doc/examples/reconst_mcsd.py:
```diff
@@ -233,16 +300,42 @@
                                 norm=False, colormap='plasma')
 
-interactive = False
+interactive = True
```
Can you put it back to `False`. Thank you

skoudoro commented on dipy/reconst/csdeconv.py:
```diff
@@ -18,10 +19,55 @@
+                        "Please use "
+                        "dipy.reconst.csdeconv.auto_response_ssst instead",
+                        since='1.2', until='1.4')
+def auto_response(gtab, data, roi_center=None, roi_radii=10, fa_thr=0.7):
```
- Why this function need to be renamed ?
- the previous signature is not respected, for backward compatibilty, all previous parameters need to be present
- we need to warn the users about this change of parameter name (`roi_radii` instead of `roi_radius`).  I do not know what will be the best way to do it. Do you have an opinion @arokem ? I propose to add `roi_radius` and if the default value change, just warn the users that the value will be overwrite by roi_radii and to use roi_radii instead.

ShreyasFadnavis replied:
@skoudoro the name needs to be changed due to different `auto_response` functions for `SSST` and `MSMT` versions I guess...

karanphil replied:
Exactly, I wanted to differentiate between `SSST` and `MSMT`. For `roi_radii`, I propose to change it to `roi_radius` in the deprecated version `auto_response`.

skoudoro replied:
ok, I'm fine, just make sure to respect the previous signature here and warn the user if there is some parameters not used anymore

karanphil replied:
Is it ok how I put the warning in the `auto_response` function?

skoudoro replied:
Yes, Great! Thank you.

karanphil commented on dipy/core/gradients.py:
```diff
@@ -674,7 +674,87 @@ def round_bvals(bvals, bmag=None):
     return b.round() * (10 ** bmag)
 
 
-def unique_bvals(bvals, bmag=None, rbvals=False):
```
It was to differentiate with unique_bvals_tolerance. Maybe I could add a deprecate unique_bvals function like I did for auto_response?

skoudoro replied:
yes

karanphil commented:
@skoudoro, I don't understand the Travis fail. I pulled the master and resolved any conflicts.

skoudoro commented:
Hi @karanphil, 

I recommend you to just add an `if` and it will be ok. something like:
``` python
from distutils.version import LooseVersion
NUMPY_1_12_PLUS = LooseVersion(numpy.__version__) >= LooseVersion('1.2.0')
rcond_value = None if NUMPY_1_12_PLUS else -1
# pass the default value
```

karanphil commented:
> Hi @karanphil,
> 
> I recommend you to just add an `if` and it will be ok. something like:
> 
> ```python
> from distutils.version import LooseVersion
> NUMPY_1_12_PLUS = LooseVersion(numpy.__version__) >= LooseVersion('1.2.0')
> rcond_value = None if NUMPY_1_12_PLUS else -1
> # pass the default value
> ```

Did I put the `import` at the right place?

skoudoro commented:
> Did I put the import at the right place?

yes, the code was just to give you an idea. I did not test it.

karanphil commented:
@skoudoro, `dipy.reconst.csdeconv.response_from_mask` has been renamed `dipy.reconst.csdeconv.response_from_mask_ssst`, should I also add a deprecated version like `auto_response`?

skoudoro commented:
> dipy.reconst.csdeconv.response_from_mask has been renamed dipy.reconst.csdeconv.response_from_mask_ssst, should I also add a deprecated version like auto_response?

yes, any renamed functions or change of parameters need a deprecation cycle

karanphil commented:
> > dipy.reconst.csdeconv.response_from_mask has been renamed dipy.reconst.csdeconv.response_from_mask_ssst, should I also add a deprecated version like auto_response?
> 
> yes, any renamed / moved functions or change of parameters need a deprecation cycle

What about `dipy.reconst.mcsd.multi_shell_fiber_response` that I moved from `dipy.sims.voxel.multi_shell_fiber_response` and greatly changed? This is the main function that had to be changed to improve the msmt-CSD results. The input parameters don't even have the same shape as before.

arokem commented:
Hey @karanphil : first of all, thanks for doing this work! Sorry for being late to the party. 

I just resolved the merge conflicts that existed here, using the web UI. If you pull this branch back onto your machine (using `git pull origin auto_response_msmt` you should have that merge conflict incorporated into this branch and you can keep going from there. 

I'll take a look a other things as well. In particular, do I understand correctly that [this comment](https://github.com/dipy/dipy/pull/2191#issuecomment-663602119) still needs to be resolved to move ahead here?

arokem commented on dipy/core/gradients.py:
```diff
@@ -9,11 +9,48 @@
+                        "Please use "
+                        "dipy.core.gradients.unique_bvals_magnitude instead",
+                        since='1.2', until='1.4')
+def unique_bvals(bvals, bmag=None, rbvals=False):
```
I feel like the changes in this module could be greatly simplified by adding a `tol` key-word argument to this function. If this key-word argument is not set (defaults to None), the function behaves as before. If `tol` is set, it takes precedence.

karanphil replied:
So I should copy my `unique_bvals_tol` into the old `unique_bvals` and use a big `if else`?

frheault replied:
To smooth the transition, can we do a mix depreciation? 1) Do your suggestion @arokem 2) but warn the user its gonna be split in 2 in Dipy 1.4? I really feel like functions doing two atomic, easy to understand, things could just be 2 functions.

Also if the decision is made to do a if/else approach in that function, @karanphil I would keep the two other, and in the unique_bvals() function, I would do it robust style. If it was my decision (which it is not, I know) I would set both params to None, the user must explicitly choose the behavior (mag or tol) by setting the parameters (likely an integer) and cannot set both.

I say this because I saw a lot of functions with mixed behavior when working on the stateful tractogram and one of the main sources of confusion was default parameters were leading to behavior # 1 but optional parameters were leading to behavior # 2. This is why I like the concept of having two functions better, short term some user will get a warning, long term everything is clearer and no functionalities is lost.

---------------
```
deprication_warning('This function will be split in two', since='1.2', until='1.4')
def unique_bvals(bvals, mag=None, tol=None rbvals=False):
    if mag is None and tol is None:
        raise
    if mag is not None and tol is not None:
         raise
    if mag is not None:
        unique_bvals_mag()
    elif tol is not None:
        unique_bvals_tol()
```

arokem commented on dipy/core/gradients.py:
```diff
@@ -674,7 +711,86 @@ def round_bvals(bvals, bmag=None):
+    bval: float or int
+        b-value to extract indices
+
+    tol: int
```
```suggestion
    tol: int, optional
```

arokem commented on dipy/core/gradients.py:
```diff
@@ -674,7 +711,86 @@ def round_bvals(bvals, bmag=None):
+
+    tol: int
+        The tolerated gap between the b-values to extract
+        and the actual b-values.
```
```suggestion
        and the actual b-values. Default: 20.
```

arokem commented on dipy/core/tests/test_gradients.py:
```diff
@@ -10,12 +10,23 @@
-                                 check_multi_b, round_bvals, unique_bvals,
+                                 check_multi_b, round_bvals, get_bval_indices,
+                                 unique_bvals_magnitude, unique_bvals_tolerance,
+                                 unique_bvals, btensor_to_bdelta)
```
Sorry : looks like my resolution of the merge conflict introduced an error:

```suggestion
                                 unique_bvals, btensor_to_bdelta,
```

arokem commented on dipy/reconst/mcsd.py:
```diff
@@ -49,21 +59,27 @@ def multi_tissue_basis(gtab, sh_order, iso_comp):
         shells : int
             Number of shells in the data
+        S0 : array (3,)
+            Signal with no diffusion weighting for each tissue compartments.
```
In what order should these be provided? Maybe better for this to be a dict: `{"wm":value, "gm": value, "csf": value}`, or three input variables: `wm_S0`, `gm_S0`, `csf_S0`

arokem replied:
The latter would also help avoid the unfortunate clash with the S0 used by `predict` that you noticed below. 

Also: I am a bit confused. Where is this variable even being used?

karanphil replied:
The order must be the same as the order of the responses. Since one could build a response with the order they want, I wanted to stay general.

karanphil replied:
S0 is never used for now. I added it for a potential futur work on `predict`. I put comments in `predict` about ideas to used a S0!=1, but I was faced with the problem of needing the volume fractions or tissue masks. This could be useful is someone wants to predict with an arbitrary S0. In my case, I only see a use for the S0=1 case.

arokem replied:
OK - now I understand. I would add this to the docstring: "This is the S0 used for predicting from a fit model" or something like that. And what do you think about breaking this out into three different parameters?

karanphil replied:
The problem is that one could want 4 tissues (for example), so I think we have to stay general.

arokem replied:
Gotcha. Yes. That makes sense. I think that expanding the docstring here would really help.

karanphil replied:
I'll do that!

arokem replied:
Awesome. Don't forget to pull from your branch before moving ahead, as I've made some changes to it!

karanphil commented:
> Hey @karanphil : first of all, thanks for doing this work! Sorry for being late to the party.
> 
> I just resolved the merge conflicts that existed here, using the web UI. If you pull this branch back onto your machine (using `git pull origin auto_response_msmt` you should have that merge conflict incorporated into this branch and you can keep going from there.
> 
> I'll take a look a other things as well. In particular, do I understand correctly that [this comment](https://github.com/dipy/dipy/pull/2191#issuecomment-663602119) still needs to be resolved to move ahead here?

Hi @arokem, thanks for your review! 

I will try to answer your questions before august 18th because I leave for vacation from the 18th to the 26th of august. I hope that you are not in a big hurry, as I would gladly get back to it afterwards. 

First of all, yes, that comment isn't resolved but I think we should let it as it is, since it is noted in api_changes.rst and the function has changed a lot and has been moved.

I will answer your other comments directly.

karanphil commented:
> OK - just a quick first pass. Sorry for breaking the tests! I think that I found the place where my merge broke things and I offer a suggested edit there.
> 
> Other than that, more generally, I wonder whether we could be a bit more gentle with the breaking changes. Some of these behaviors are changes in API that would not matter for most users (who are using the functions).
> 
> This is to say that I wonder whether it makes sense to leave `auto_response` and change it to have the functionality now implemented in `auto_response_ssst`. This is essentially a fix, so I think that users should be (mostly) pleased with the changes.
> 
> I'll continue to look at this, but thought I would offer a first volley of comments first (with a fix for my breakage. Apologies again!)
> 
> And thanks for this really nice important work!

About `auto_response`, it is not really a fix, I just simplified the function by splitting it in two distinct functions `mask_for_response_ssst` and `response_from_mask_ssst`. I renamed it to avoid confusion with the new function `auto_response_msmt`. So that is why I put a deprecated version of it at thte beginning, as the way to use it hasn't really changed much.

arokem commented:
> About auto_response, it is not really a fix, I just simplified the function by splitting it in two distinct functions mask_for_response_ssst and response_from_mask_ssst. I renamed it to avoid confusion with the new function auto_response_msmt. So that is why I put a deprecated version of it at thte beginning, as the way to use it hasn't really changed much.

I would be fine having `csd.auto_response` and `msmt.auto_response` as two functions with the same name that differ by the namespace they belong to. If that confuses people, we can write examples that have `from msmt import auto_response as auto_response_msmt`. But if we stick to this naming by module, all our examples remain as they were and previous code doesn't have to be changed. And you don't need to have these deprecation warnings. I think that would be better -- but open to arguments against this.

arokem commented:
I just merged the fix to the broken test. I hope that undoes the breakage I introduced yesterday.

arokem commented on dipy/core/tests/test_gradients.py:
```diff
@@ -10,12 +10,23 @@
                                  WATER_GYROMAGNETIC_RATIO,
                                  reorient_bvecs, generate_bvecs,
-                                 check_multi_b, round_bvals, unique_bvals,
+                                 check_multi_b, round_bvals, get_bval_indices,
+                                 unique_bvals_magnitude, unique_bvals_tolerance,
+                                 unique_bvals, btensor_to_bdelta,
```
OK - trying this one more time:

```suggestion
                                 check_multi_b, round_bvals, get_bval_indices,
                                 unique_bvals_magnitude, unique_bvals_tolerance,
                                 unique_bvals, 
```

karanphil commented:
> > About auto_response, it is not really a fix, I just simplified the function by splitting it in two distinct functions mask_for_response_ssst and response_from_mask_ssst. I renamed it to avoid confusion with the new function auto_response_msmt. So that is why I put a deprecated version of it at thte beginning, as the way to use it hasn't really changed much.
> 
> I would be fine having `csd.auto_response` and `msmt.auto_response` as two functions with the same name that differ by the namespace they belong to. If that confuses people, we can write examples that have `from msmt import auto_response as auto_response_msmt`. But if we stick to this naming by module, all our examples remain as they were and previous code doesn't have to be changed. And you don't need to have these deprecation warnings. I think that would be better -- but open to arguments against this.

I think I would need to have the deprecation warnings since the signature of the function has changed. Am I right? Also, just to be clear, there are also the following functions: `response_from_mask_ssst` `mask_for_response_ssst` `response_from_mask_msmt` `mask_for_response_msmt`.

arokem commented:
> I think I would need to have the deprecation warnings since the signature of the function has changed. Am I right? Also, just to be clear, there are also the following functions: response_from_mask_ssst mask_for_response_ssst response_from_mask_msmt mask_for_response_msmt.

The default API actually doesn't change, right? Although the underlying behavior changes. So, yes: you need to warn users (e.g., in `api_changes.rst` that the function behavior has changed), but arguably this is an improvement, right? That's why you want to deprecate this function in the first place. And I think it would be fine to add `csd.response_from_mask`, `msmt.response_from_mask`, `csd.mask_for_response`, and `msmt.mask_for_response`. They have similar APIs across the modules, right? So a downstream user might automatically choose between CSD and MSMT based on the data and then use the same exact code to do the analysis. I find that appealing.

karanphil commented:
> > I think I would need to have the deprecation warnings since the signature of the function has changed. Am I right? Also, just to be clear, there are also the following functions: response_from_mask_ssst mask_for_response_ssst response_from_mask_msmt mask_for_response_msmt.
> 
> The default API actually doesn't change, right? Although the underlying behavior changes. So, yes: you need to warn users (e.g., in `api_changes.rst` that the function behavior has changed), but arguably this is an improvement, right? That's why you want to deprecate this function in the first place. And I think it would be fine to add `csd.response_from_mask`, `msmt.response_from_mask`, `csd.mask_for_response`, and `msmt.mask_for_response`. They have similar APIs across the modules, right? So a downstream user might automatically choose between CSD and MSMT based on the data and then use the same exact code to do the analysis. I find that appealing.

The positional arguments of `auto_response` are the same, but the optionnal arguments have changed. So if someone was using for instance `fa_callable=...`, it would not work. Is this a problem?

Also, yes, the functions have similar APIs across the modules but the MSMT ones are more complicated, including three tissues. So not exactly the same code unfortunately. 

I can remove `_ssst` and `_msmt` from everywhere if you want! Maybe it would be a good opportunity to find a better name for the `mcsd.py` module? It can be confusing to have `csdeconv.py` and than `mcsd.py`.

arokem commented:
>The positional arguments of auto_response are the same, but the optionnal arguments have changed. So if someone was using for instance fa_callable=..., it would not work. Is this a problem?

Yeah - might be. One way about this is to raise an informative error if someone uses these deprecated arguments. Something like: "As of version 1.2, the API of `auto_response` has changed. The keyword argument `fa_callable` is no longer valid". I wonder whether others think this is too aggressive towards our users.

> Also, yes, the functions have similar APIs across the modules but the MSMT ones are more complicated, including three tissues. So not exactly the same code unfortunately.
> I can remove _ssst and _msmt from everywhere if you want! Maybe it would be a good opportunity to find a better name for the mcsd.py module? It can be confusing to have csdeconv.py and than mcsd.py.

Yeah - I would prefer same names in the two modules. About `mcsd.py` what did you have in mind? `csdmsmt.py`? Or something else?

skoudoro commented:
>  I wonder whether others think this is too aggressive towards our users.

I do not think it is too aggressive. 

> Yeah - I would prefer same names in the two modules

+1

karanphil commented:
>> Yeah - might be. One way about this is to raise an informative error if someone uses these deprecated arguments. Something like: "As of version 1.2, the API of `auto_response` has changed. The keyword argument `fa_callable` is no longer valid". I wonder whether others think this is too aggressive towards our users.
>
>I am not sure about how to do this properly. Also I leave tomorrow for a week vacation. I think your idea is good and I would gladly look at it more when I come back! If you are in a hurry, could you do it?
> 
> >Yeah - I would prefer same names in the two modules. About `mcsd.py` what did you have in mind? `csdmsmt.py`? Or something else?
>
>Not sure... at least `mcsdeconv.py` to fit with `csdeconv.py` and then the `m` stands for `multi`? Maybe `msmtcsdeconv.py`, but this is quit confusing, so `msmt_csd.py`? These are just ideas.

Garyfallidis commented:
@karanphil thank you for this work. Can you fix conflict or rebase?

karanphil commented:
> @karanphil thank you for this work. Can you fix conflict or rebase?

Yes, but I still have to address some changes @arokem requested.

Garyfallidis commented:
Agreed @karanphil

Garyfallidis commented on doc/examples/contextual_enhancement.py:
```diff
@@ -117,15 +117,15 @@
 
 # Perform CSD on the original data + noise
-response, ratio = auto_response(gtab, data_noisy, roi_radius=10, fa_thr=0.7)
+response, ratio = auto_response_ssst(gtab, data_noisy, roi_radii=10, fa_thr=0.7)
```
Can you please explain why this is a good change here?

Garyfallidis commented on doc/examples/reconst_csd_parallel.py:
```diff
@@ -31,9 +31,9 @@
+from dipy.reconst.csdeconv import auto_response_ssst
 
-response, ratio = auto_response(gtab, maskdata, roi_radius=10, fa_thr=0.7)
+response, ratio = auto_response_ssst(gtab, maskdata, roi_radii=10, fa_thr=0.7)
```
I am not sure why to use ssst here. Can we not have autoresponse with an option to use ``single_shell`` or ``multi_shell``? Not sure why to change the name of the function which introduces compatibility issues.

karanphil replied:
As discussed with @arokem, we will have an `auto_response` function in `reconst.csdeconv.py` and one in `reconst.mcsd.py`. I just came back from vacation, I will change this new monday, sorry for the delay.

frheault replied:
@arokem @Garyfallidis While I appreciate having a simple name in the right module, I think using the generic name `auto_response` is what led us to this PR in the first place. The naming convention was not thought in advance and led to a confusing software design. What if someone codes a single-shell-multi-tissues (ssmt) algorithm? What if someone codes Chantal Tax recursive fiber response function (this one would go in CSD too)? What if someone wants to add a new (unknown) algorithm for fiber response in MSMT (like in mrtrix, using 5ttgen input) ?

 All of the above examples would need/want a auto-response that evaluates the mask as they see fit, then compute the response, maybe even post-processing, and then return it.

This whole PR is refactoring functions that had an appropriate name at the time, but didn't age well and it took someone a lot of motivation to "fix" it retroactively to introduce a new kind of CSD and keep it well organized. I think it would be a mistake to limit ourself to one `auto_response` per module since we do not know what is ahead of use.

arokem replied:
OK - on quite a bit of further thought, I think that I am OK with the PR as is. 

Given two reviewers approve (me and @frheault), maybe we can include this in 1.2 after all?

I don't see anything outstanding, except for this issue, but I might be missing something?

Garyfallidis replied:
@frheault what you say makes sense. Agreed.

Garyfallidis commented:
See also my comments before rebasing.

frheault commented:
@skoudoro What is the timeline for Dipy 1.3? This is 99% done, and the test are passing, it would be sad to wait to next June for that fix?
Are you aiming for a smaller 1.3 release before christmas?

skoudoro commented:
Hi @frheault,

You can see the timeline of the next releases on this page: https://github.com/dipy/dipy/milestones. 

> EDIT: For me, this is ready to merge as-is.

ok, however, it seems that all concerns are not resolved yet. @arokem and @Garyfallidis will give you some feedback soon

frheault commented:
@karanphil Put that PR into gear for 1.3 in October ;)

If my argument for the function name and behavior 'clashes' with Ariel and Elef, it would be your call to decide who to listen to in order to speed up the potential merge before 1.2 :P

skoudoro commented:
> If my argument for the function name and behavior 'clashes' with Ariel and Elef, it would be your call to decide who to listen to in order to speed up the potential merge before 1.2 :P

ehehe 😄 ,  we are not at this point yet, I think there is no need to rush to merge and I can not delay more 1.2.  So 1.3 is a good compromise.

But thank you for both for the reactivity and for pinging us!

frheault commented:
@skoudoro Fair enough, it is already a miracle that this PR is going so well despite its complexity, the fact that its summer vacations and a global pandemic!

@karanphil You heard the man, be sure its the first PR ready to be merged after 1.2 is release

karanphil commented:
Hi @skoudoro,

@arokem seems to be OK with the PR. Is it too late for 1.2?

Also, I see trouble with two checks, is it my fault?

Garyfallidis commented:
Merging. Congrats @karanphil you made it to 1.2.
